src/util/hw_topo/mpir_hw_topo.c

```diff
@@ -0,0 +1,168 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ */
+
+#include "mpiimpl.h"
+#include "mpir_hw_topo.h"
+
+#ifdef HAVE_HWLOC
+#include "hwloc.h"
+#endif
+
+#ifdef HAVE_NETLOC
+#include "netloc.h"
+#endif
+
+/*
+ * Hardwre topology attributes
+ */
+static struct MPIR_Hardware_topology_attr {
+#ifdef HAVE_HWLOC
+    hwloc_topology_t hwloc_topology;
+    hwloc_cpuset_t bindset;
+#endif
+    int bindset_is_valid;
+#ifdef HAVE_NETLOC
+    netloc_topology_t netloc_topology;
+    MPIR_Netloc_network_attributes network_attr;
+#endif
+} MPIR_Hardware_topology_attr = {
+0};
+
+/*
+ * Function prototype definitions
+ */
+#ifdef HAVE_HWLOC
+static int MPIR_HWLOC_get_covering_nodeobj(MPIR_Node_hw_obj_type obj_type, int *size,
+                                           int **obj_ids);
+static int MPIR_HWLOC_mbind(void *addr, size_t len, MPIR_Node_hw_obj_type mem_obj_type,
+                            int *obj_ids, MPIR_Membind_policy policy, MPIR_Membind_flags flags);
+#else
+static int MPIR_GEN_get_covering_nodeobj(MPIR_Node_hw_obj_type obj_type, int *size, int **obj_ids);
+static int MPIR_GEN_mbind(void *addr, size_t len, MPIR_Node_hw_obj_type mem_obj_type, int *obj_ids,
+                          MPIR_Membind_policy policy, MPIR_Membind_flags flags);
+#endif
```

Comment:
Can the function name be kept the same and two different implementations be provided with or without hwloc?

Reply:
Sure, just wanted to have a distinction between the names to separate the two. So in principle you could, theoretically, have another topology backend (e.g., `MPIR_XXX_`) and still have a general common function to fall back to if neither `MPIR_HWLOC` or `MPIR_XXX` is available. But I guess since we don't have additional hardware topology libraries support this is superfluous.

Reply:
I understand, but I guess my point was to have just one function name, which would map to different backends depend on what we configure mpich with.